x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/PasswordAndEnrollmentInitialNode.java

```diff
@@ -0,0 +1,239 @@
+        KeyStoreWrapper.SEED_SETTING);
+    private static final String elasticUser = ElasticUser.NAME;
+
+    private SecureString password;
+    private String token;
+    private String fingerprint;
+
+    // Protected for testing
+    protected SecureString getPassword() {
+        return password;
+    }
+
+    protected String getToken() {
+        return token;
+    }
+
+    protected String getFingerprint() {
+        return fingerprint;
+    }
+
+    protected CommandLineHttpClient getClient(Environment env) {
+        return new CommandLineHttpClient(env);
+    }
+
+    protected CreateEnrollmentToken getCreateEnrollmentToken(Environment env) throws Exception{
+        return new CreateEnrollmentToken(env);
+    }
+
+
+    PasswordAndEnrollmentInitialNode (){
+        super("Set elastic password and generate enrollment token for initial node");
+        parser.allowsUnrecognizedOptions();
+    }
+
+    public static void main(String[] args) throws Exception {
+        exit(new PasswordAndEnrollmentInitialNode().main(args, Terminal.DEFAULT));
+    }
+
+    @Override
+    protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        if (options.nonOptionArguments().contains("--explicitly-acknowledge-execution") == false) {
+            throw new UserException(ExitCodes.NOOP, "This command is not intended for end users");
+        }
+        if (env.settings().hasValue(XPackSettings.ENROLLMENT_ENABLED.getKey()) && false ==
+            XPackSettings.ENROLLMENT_ENABLED.get(env.settings())) {
+            throw new UserException(ExitCodes.NOOP, "Enrollment is explicitly disabled.");
+        }
+        CommandLineHttpClient client = getClient(env);
+        CreateEnrollmentToken cet = getCreateEnrollmentToken(env);
+
+        // Try 5 times and rethrow the last exception from checkClusterHealth
+        for (int retry = 5; ; retry--) {
+            try {
+                checkClusterHealth(env, client);
+                break;
+            } catch (Exception e) {
+                if (retry == 1) {
+                    throw e;
+                }
+                Thread.sleep(1000);
+            }
+        }
+        if (Strings.isNullOrEmpty(bootstrapPassword.get(env.settings()).toString())) {
+            changeElasticUserPassword(env, terminal, client);
+        } else {
+            password = bootstrapPassword.get(env.settings());
+        }
+        token = cet.createKibanaEnrollmentToken(elasticUser, password);
+        Map<String, String> infoNode = getDecoded(token);
+        fingerprint = infoNode.get("fgr");
+
+        terminal.println("'elastic' user password: " + password);
+        terminal.println("enrollment token: " + token);
+        terminal.println("CA fingerprint: " + fingerprint);
+    }
+
+    protected void checkClusterHealth(Environment env, CommandLineHttpClient client) throws Exception {
```

Comment:
Please merge master and factor out this method, because it used in other places. I would suggest we add a method to the `CommandLineHttpClient`, but I would be satisfied if we avoid coding this everytime we need it.